File: Excepciones2/lector_archivo.py
import sys


class NombreArchivoIncorrectoError(Exception):
    
    def __init__(self, mensaje_error):
        super().__init__(self, mensaje_error)


class ExtensionArchivoIncorrectaError(Exception):

    def __init__(self, mensaje_error):
        super().__init__(self, mensaje_error)


class LectorArchivo:


    def __contiene_extension(nombre_archivo):
        if (".txt" in nombre_archivo or  ".doc" in nombre_archivo):
            return (True)
        return (False)


    def __es_correcto_nombre_archivo(nombre_archivo):
        if (nombre_archivo == "nombreArchivoErroneo.txt"):
            return (False)
        else:
            return (True)


    def primera_linea(nombre_archivo):
        try:
            archivo = open(nombre_archivo, "r")
            linea = archivo.readline()
            if (linea != ''):
                return linea
            else:
                raise EOFError("El archivo no se puede leer: " + nombre_archivo);
        except FileNotFoundError as err:
            if (not LectorArchivo.__es_correcto_nombre_archivo(nombre_archivo)):
                raise NombreArchivoIncorrectoError("Nombre de archivo incorrecto: " + nombre_archivo)
            else:
                raise err

        except EOFError as err:
            if (not LectorArchivo.__contiene_extension(nombre_archivo)):
                raise ExtensionArchivoIncorrectaError("Extensión de archivo incorrecta: " + nombre_archivo)
            else:
                raise err


linea = LectorArchivo.primera_linea(sys.argv[1])
print (linea)

File: Excepciones2/LectorArchivo.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class LectorArchivo {

    private static boolean contieneExtension(String nombreArchivo) {
        if (nombreArchivo.contains(".txt") || nombreArchivo.contains(".doc"))
            return (true);
        return (false);
    }

    private static boolean esCorrectoNombreArchivo(String nombreArchivo) {
        if (nombreArchivo.equals("nombreArchivoErroneo.txt"))
            return (false);
        else
            return (true);
    }
    
    public static String primeraLinea(String nombreArchivo)
            throws NombreArchivoIncorrectoException, IllegalArgumentException, ExtensionArchivoIncorrectaException, FileNotFoundException {
        try {
            Scanner archivo = new Scanner(new File(nombreArchivo));
            if (archivo.hasNextLine()) {
                return archivo.nextLine();
            } else {
                throw new IllegalArgumentException(nombreArchivo
                        + " (El archivo no se puede leer)");
            }
        } catch (FileNotFoundException err) {
            if (!esCorrectoNombreArchivo(nombreArchivo)) {
                throw new NombreArchivoIncorrectoException(nombreArchivo
                        + " (Seleccionar otro archivo)");
            } else {
                throw err;
            }

        } catch (IllegalArgumentException err) {
            if (!contieneExtension(nombreArchivo)) {
                throw new ExtensionArchivoIncorrectaException(nombreArchivo
                        + " (Sólo leo archivos de texto)" );
            } else {
                throw err;
            }
        }
    }

    public static void main(String[] args) {
        try {
            System.out.println(primeraLinea(args[0]));
        } catch (ExtensionArchivoIncorrectaException | IllegalArgumentException
                 | NombreArchivoIncorrectoException | FileNotFoundException err) {
            System.out.println(err);
        }
    }

}
